pkg/scanners/clairv4/convert.go

```diff
@@ -0,0 +1,187 @@
+	// to make it easier to set the (*claircore.Manifest).Hash.
+	imgDigest, err := claircore.ParseDigest(imageutils.GetSHA(image))
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing image digest")
+	}
+	manifest := &claircore.Manifest{
+		Hash: imgDigest,
+	}
+
+	client := registry.HTTPClient()
+	remote := image.GetName().GetRemote()
+	for _, layerSHA := range image.GetMetadata().GetLayerShas() {
+		layerDigest, err := claircore.ParseDigest(layerSHA)
+		if err != nil {
+			return nil, errors.Wrap(err, "parsing image layer digest")
+		}
+
+		uri, header, err := fetchLayerURIAndHeader(client, cfg.URL, remote, layerDigest.String())
+		if err != nil {
+			return nil, err
+		}
+
+		// Layers needs to be ordered from base -> top layer, which is the same as how the metadata sorts them.
+		manifest.Layers = append(manifest.Layers, &claircore.Layer{
+			Hash:    layerDigest,
+			URI:     uri,
+			Headers: header,
+		})
+	}
+
+	return manifest, nil
+}
+
+// fetchLayerURIAndHeader is based on the clairctl v4.5.0 implementation for creating manifests:
+// https://github.com/quay/clair/blob/v4.5.0/cmd/clairctl/manifest.go#L76.
+// It is assumed the given client handles auth.
+func fetchLayerURIAndHeader(client *http.Client, url, repository, digest string) (string, http.Header, error) {
+	path := fmt.Sprintf("/v2/%s/blobs/%s", repository, digest)
+	req, err := http.NewRequest(http.MethodGet, url+path, nil)
+	if err != nil {
+		return "", nil, errors.Wrap(err, "creating image layer pull request")
+	}
+	// We don't actually want the layer. We just want the headers.
+	req.Header.Add("Range", "bytes=0-0")
+	res, err := client.Do(req)
+	if err != nil {
+		return "", nil, errors.Wrap(err, "fetching image layer")
+	}
+	utils.IgnoreError(res.Body.Close)
+
+	res.Request.Header.Del("User-Agent")
+	res.Request.Header.Del("Range")
+
+	return res.Request.URL.String(), res.Request.Header, nil
+}
+
+// imageScan converts the given report to an image scan.
+func imageScan(report *claircore.VulnerabilityReport) *storage.ImageScan {
+	scan := &storage.ImageScan{
+		ScanTime:        gogotypes.TimestampNow(),
+		Components:      components(report),
+		OperatingSystem: os(report),
+	}
+
+	if scan.GetOperatingSystem() == "unknown" {
+		scan.Notes = append(scan.Notes, storage.ImageScan_OS_UNAVAILABLE)
+	}
+
+	return scan
+}
+
+func components(report *claircore.VulnerabilityReport) []*storage.EmbeddedImageScanComponent {
+	components := make([]*storage.EmbeddedImageScanComponent, 0, len(report.PackageVulnerabilities))
+	for id, pkg := range report.Packages {
+		vulnIDs := report.PackageVulnerabilities[id]
+		component := &storage.EmbeddedImageScanComponent{
+			Name:    pkg.Name,
+			Version: pkg.Version,
+			Vulns:   vulnerabilities(report.Vulnerabilities, vulnIDs),
+		}
+
+		components = append(components, component)
+	}
+
+	return components
+}
+
+// vulnerabilities returns a list of vulnerabilities from the given map indexed by the given ids.
```

Comment:
so the map idx to vulns map?

Reply:
Right. We get the slice of `ids` from `PackageVulnerabilities`. Both that and `Vulnerabilities` which is the map of vulns are defined below:

```
// all discovered vulnerabilities affecting this manifest
Vulnerabilities map[string]*Vulnerability `json:"vulnerabilities"`
// a lookup table associating package ids with 1 or more vulnerability ids. keyed by package id
PackageVulnerabilities map[string][]string `json:"package_vulnerabilities"`
```